### DISCUSSION TITLE
DatePicker behaves different on first and second selection for dependant date field with max date

### DISCUSSION THREAD

**Discussion by citroid** [filamentphp/filament#5506] [Help]
Hi,
I have built a form for my resource which contains two dependant date fields (with DatePicker as input element).
The first date is an event date (required field) and the second date is an optional date for a day of arrival.
If set, the arrival date should be restricted to a max date of one day before the event date.
However, the behaviour I'm obeserving is kind of strange and I am not sure if my implementation has a flaw or if I came across a bug of the DatePicker component.
The approach is as follows:

1. I select a value for event date (e.g. 26.01.2023)
2. When I select the arrival date **for the first time** the maximum allowed value is **two days before** event date (e.g. 24.01.2023)
3. After selecting an arrival date for the first time, I try to pick a different date and **on the second time** entering the DatePicker it **now offers me the day before** event date as maximum (e.g. 25.01.2023)
 
I don't why this behaves differently on first and second time opening the DatePicker.
Is this a bug or did I make a mistake here?

Here's the relevant part of my code:
```
public static function form(Form $form): Form
 {
        return $form
            ->schema([
		  // ...
		  DatePicker::make('event_date')
                    ->label('Event date')
                    ->columnSpan(6)
                    ->displayFormat('d.m.Y')
                    ->closeOnDateSelection()
                    ->minDate(now())
                    ->reactive()
                    ->required(),
                  DatePicker::make('arrival_date')
                    ->label('Arrival date')
                    ->columnSpan(6)
                    ->displayFormat('d.m.Y')
                    ->closeOnDateSelection()
                    ->minDate(now())
                    ->maxDate(function (Closure $get) {
                        // Calculate max of arrival date depending on PE date
                        $eventDate = $get('event_date');
                        if ($eventDate != null) {
                            return Carbon::parse($eventDate)->subDays(1);
                        }
                        return null;
                      }),
		      // ...
		]);
	}
```

**Comment by danharrin** [comment]
Maybe it's because of timezones or something? 🤔

**Reply by citroid** [reply]
Hi Dan,
Thank you for the response! 
I did not explicitly set a timezone for the form fields in my code (only in config/app.php it is set to 'Europe/Berlin').
To further narrow it down I simplified my maxDate constraint to a constant date string:
```
DatePicker::make('arrival_date')
// ...
->maxDate('2023-01-30')
```
And just as before I could make the same observation. On my first select I am restricted to 2023-01-29 and on the next attempt I may also choose 2023-01-30.
So apparently it has nothing to do with the two fields being dependent, but I'm still confused what causes this behaviour.
Is there a good spot where I could place a debug output to make the inner workings more transparent here?

**Reply by danharrin** [reply]
Without a reproduction repo we can't help you debug

**Reply by citroid** [reply]
I tried to provide a minimal example on Github, but it behaves differently than I expected (though I applied the same logic):
https://github.com/citroid/filamentphp-datepicker

The form in app\Http\Livewire\DemoComponent.php contains the following code in getFormSchema():

```
DatePicker::make('event_date')
                ->label('Event Date')
                ->displayFormat('d.m.Y')
                ->closeOnDateSelection()
                ->minDate(now())
                ->reactive()
                ->required(),
DatePicker::make('arrival_date')
                ->label('Arrival Date')
                ->displayFormat('d.m.Y')
                ->closeOnDateSelection()
                ->minDate(now())
                ->maxDate(function (callable $get) {
                    // Calculate max of arrival date depending on event date
                    $eventDate = $get('event_date');
                    if ($eventDate == null) {
                        Log::info("Retrieved value of event_date is null.");
                        return null;
                    }
                    else {
                        Log::info("Retrieved value of event_date: " . Carbon::parse($eventDate)->toDateString());
                        Log::info("Set maxDate for 'arrival_date' to: " . Carbon::parse($eventDate)->subDays(1)->toDateString());
                        return Carbon::parse($eventDate)->subDays(1);
                    }
                }),
```

It can be tested locally by visiting this URL: http://127.0.0.1:8000/demo

Contrary to what I would expect the function in maxDate() (on field "arrival_date") seems to be called only once when the page is loaded for the first time and $get('event_date') returns null (as can be seen in laravel.log). This explains why it has no effect, but not why it is not called again when the value of "event_date" changes (since this field is marked as reactive).

Did I make a mistake here?

**Reply by danharrin** [reply]
No, unfortunately the max date cannot be modified after the initial page load. Sorry

**Reply by citroid** [reply]
That's weird, because in my original example maxDate() can definitely be changed after the initial page load.
The only difference is, that this is done within a Filament resource's form() method while my simplified example tried to recreate the same scenario using only filament/forms...
I'll try to provide a minimal example with a Filament resource which also shows the same problem (different date selections on first and second try).

**Reply by danharrin** [reply]
Hm maybe I was wrong about it not being possible, but I really do need a minimal reproduction. It's not very easy to debug when there's other fields involved

**Reply by citroid** [reply]
Yes, I understand that. I'll try it with a Filament resource form having only those two dependant DatePicker fields to keep it as simple as possible.

**Reply by citroid** [reply]
Finally I've been successful in reproducing the issue in a Github repo:
https://github.com/citroid/filamentphp-datepicker2

There is only one Filament resource (EventResource) and the corresponding form features the same constellation of "event_date" and "arrival_date" as posted above. I'm also observing the same behaviour that I have posted originally: the first time I can't select the correct max date from the dependant field, but from the second time on it works fine. The laravel.log also shows that max date should have been set correctly.

The only things you should have to change is the database path in the .env file (I used sqlite) and you would have to create a Filament user of course.

Please let me know if I can help with more information on this issue.

**Reply by danharrin** [reply]
Could you open an issue so we can track this bug easier?

**Reply by citroid** [reply]
Yes, I just opened an [issue](https://github.com/filamentphp/filament/issues/5808) for this.

**Reply by danharrin** [reply]
Thanks, can't guarantee when I'll be able to work on this but at least you can now track any progress
